### PR TITLE
Add DevSecOps on Azure using Terraform and Python

### PR DESCRIPTION
**Pull request by RaphHaddad**

### CODE REVIEW

File: content/post/devsecops-azure-terraform-python.md [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+This post will demonstrate how to create a segregated Azure Cloud environment (a
+resource group and service
+connection from Azure DevOps) that can be used by development teams to provision their own infrastructure and
+deploy their compiled code. All in an automated and secure manner.
+
+Teams can then setup their own release pipelines using the provisioned resource
+group and service connection. The service connection created will only have
+permission to a single resource group within Azure.
+
+The reference code for this post is available on this [GitHub repository](https://github.com/RaphHaddad/devsecops-azure-terraform-python).
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:17]: This feels out of order to me. You talk about what the reader will get out of it, which is good (ideally a heading saying something like "what you will learn" or similar. I'd place the "The reference for this code ...." a bit lower and group the "Specifically, this post will...." together with the above or even put the link to the reference coe right at the top.

File: content/post/devsecops-azure-terraform-python.md [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+The first step is authenticating to an Azure DevOps instance and an Azure Cloud
+Subscription. To authenticate to Azure Cloud the [Azure
+CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) is needed.
+
+```cmd
+% az login
+The default web browser has been opened at https://login.microsoftonline.com/common/oauth2/authorize. Please continue the login in the web browser. If no web browser is available or if the web browser fails to open, use device code flow with `az login --use-device-code`.
+```
+
+To authenticate into Azure DevOps. A [Personal Access Token (PAT)](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page) is needed with
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:47]: I think you meant a comma after "Azure Devops" not a full stop.

File: content/post/devsecops-azure-terraform-python.md [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+provider](https://registry.terraform.io/providers/hashicorp/azurerm/latest) and
+an [Azure
+DevOps](https://registry.terraform.io/providers/microsoft/azuredevops/latest/docs)
+provider. Both of these will be utilised in order to create a resource group
+with a service connection.
+
+The first step is authenticating to an Azure DevOps instance and an Azure Cloud
+Subscription. To authenticate to Azure Cloud the [Azure
+CLI](https://docs.microsoft.com/en-us/cli/azure/install-azure-cli) is needed.
+
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:41]: Azure CLI is one method. You can use powershell as well. I personally prefer, "This post will use Azure CLI as it is a very simple way of deploying to Azure cloud. To authenticate with Azure CLI you......" . Technically you do not need Azure CLI to authenticate, its just one method.

File: content/post/devsecops-azure-terraform-python.md [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+% export AZDO_ORG_SERVICE_URL=https://dev.azure.com/{a-secure-org}
+```
+
+## Initialising Terraform
+
+Terraform relies on a state file to perform resource deployments, for the
+purposes of this post, local system state will be used. For production
+work-loads [remote state](https://www.terraform.io/docs/language/state/index.html) should be used.
+
+The Python script [setup-environment.py](https://github.com/RaphHaddad/devsecops-azure-terraform-python/blob/107fe97dbd6674d1200436922b0b7d87328a480b/setup-environment.py#L13-L22), initialises Terraform state *dynamically* by assigning a
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:62]: Any reason you are using python? It is not bad, but explaining why in this instance might be useful, perhaps a little earlier in the piece

File: content/post/devsecops-azure-terraform-python.md [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+`.terraform-state-a-new-resource-group` and
+`.terraform-state-a-new-resource-group-2`. Both of which have a `tfstate` file.
+
+This pattern can be achieved for any Terraform remote backend and will allow the
+executor to use the same Terraform templates to create multiple instances of the
+resources declared in the Terraform template.
+*Note* it is important to treat Terraform state files as sensitive data
+(passwords, tokens). See this [Terraform page](https://www.terraform.io/docs/language/state/sensitive-data.html?_ga=2.252783049.1925722553.1613943599-1447805869.1611735730#recommendations) for more information.
+
+## Creating the Service Principle
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:99]: "Principal" not principle.

I am a "Principal consultant" vs "I have a principle of never eating Yoghurt". Different meanings.

Comment (glav) [reply content/post/devsecops-azure-terraform-python.md:99]: See https://adamtheautomator.com/azure-service-principal/

Comment (glav) [reply content/post/devsecops-azure-terraform-python.md:99]: I see you have used that spelling quite a bit - prolly do a big search and replace

File: content/post/devsecops-azure-terraform-python.md [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  special          = true
+  override_special = "_%@"
+}
+```
+
+## Creating the Resource Group
+
+The new resource group needs to be declared with the Terraform resource type `azurerm_role_assignment`
+with `Contributor` access given to the previously declared [service principle](#creating-the-service-principle). The
+snippet of the code is show below and the full code is available [on GitHub](https://github.com/RaphHaddad/devsecops-azure-terraform-python/blob/107fe97dbd6674d1200436922b0b7d87328a480b/azure-cloud.tf#L1-L10)
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:131]: shown

File: content/post/devsecops-azure-terraform-python.md [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+}
+```
+
+## Conclusion
+
+The above setup creates a resource group and a service principle with
+`Contributor` access to that resource group. It also creates a service
+connection from Azure DevOps that uses the provisioned service principle that
+can then be used in
+release pipelines.
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:177]: While this is technically true, what benefit is it. As in, you can now not only automate your resource creation but also automate the creation of your automated pipeline to create resources which is another level of automation again and allows good management, particularly where automation and governance is strictly required or something like that

File: content/post/devsecops-azure-terraform-python.md (hunk 2 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:47]: will fix
```suggestion
To authenticate into Azure DevOps, A [Personal Access Token (PAT)](https://docs.microsoft.com/en-us/azure/devops/organizations/accounts/use-personal-access-tokens-to-authenticate?view=azure-devops&tabs=preview-page) is needed with
```

File: content/post/devsecops-azure-terraform-python.md (hunk 4 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:62]: because Microsoft likes Linux now and Python is included natively in linux. I will re-arrange.

File: content/post/devsecops-azure-terraform-python.md (hunk 5 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:99]: Wow big error...will fix thanks Glav

File: content/post/devsecops-azure-terraform-python.md (hunk 6 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:131]: Will fix
```suggestion
snippet of the code is shown below and the full code is available [on GitHub](https://github.com/RaphHaddad/devsecops-azure-terraform-python/blob/107fe97dbd6674d1200436922b0b7d87328a480b/azure-cloud.tf#L1-L10)
```

File: content/post/devsecops-azure-terraform-python.md (hunk 1 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:17]: Will fix

File: content/post/devsecops-azure-terraform-python.md (hunk 3 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:41]: Will fix

File: content/post/devsecops-azure-terraform-python.md (hunk 5 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md]: Fixed

File: content/post/devsecops-azure-terraform-python.md (hunk 3 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:40]: @glav, Terraform's documentation only references authentication via the Azure CLI. Powershell is not mentioned. I will add a link to Terraform's documentation for the readers: https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/azure_cli

Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:40]: Resolving with commit 6c1b65c92c437bc9567cc99ed486cd3847d3dfa3

File: content/post/devsecops-azure-terraform-python.md (hunk 1 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md]: resolving with commit 6be821095e86f3a686b2f9e733addc2b3b94bf3f

File: content/post/devsecops-azure-terraform-python.md (hunk 7 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md]: Resolved with 2ec93a1503d245799acb09db2b7f622f7bed60d5

File: content/post/devsecops-azure-terraform-python.md (hunk 4 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:64]: Actually @glav, I'm not going to justify the usage of Python, because I'd feel I have to justify Terraform. Perhaps a justification of technology choice of these environments is for another blog post (I don't think I will do it justice).

**Comment by RaphHaddad** [comment]
@glav, please review again when you are able.

File: content/post/devsecops-azure-terraform-python.md [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+
+```cmd
+% export AZDO_PERSONAL_ACCESS_TOKEN={a-secure-access-token}
+% export AZDO_ORG_SERVICE_URL=https://dev.azure.com/{a-secure-org}
+```
+
+## Initialising Terraform
+
+Terraform relies on a state file to perform resource deployments, for the
+purposes of this post, local system state will be used. For production
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:61]: Just a slight grammatical change. The 2 comma's read a little wrong. I'd reword:
"Terraform relies on a state file to perform resource deployments. For the purposes of this post and to keep things simple, local system state will be used. "

File: content/post/devsecops-azure-terraform-python.md [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+Apply complete! Resources: 7 added, 0 changed, 0 destroyed.
+```
+
+The result would be a resource group `an-example-resource-group` provisioned on Azure Cloud with
+the service principal `an-example-resource-group-spn` having `Contributor`
+access to that resource group.
+
+![Resource Groups Provisioned with Terraform](/images/resource-group-terraform.png)
+
+As well as, a service connection from Azure DevOps into Azure Cloud that can be
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:200]: Remove the comma. A comma acts as a pause, and causes the flow of the paragraph to feel wrong.

File: content/post/devsecops-azure-terraform-python.md [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+used by release pipelines.
+
+![Resource Groups Provisioned with Terraform](/images/service-connection-terraform.png)
+
+Finally, thank you for reading this post. I hope you got out of it something.
```
Comment (glav) [inline content/post/devsecops-azure-terraform-python.md:205]: Reword "I hope you got out of it something." to 
"I hope you got something out of it"
or
"I hope this post has been beneficial for you".

**Review by glav — CHANGES_REQUESTED** [review]

File: content/post/devsecops-azure-terraform-python.md (hunk 10 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:205]: Will fix.
```suggestion
Finally, thank you for reading this post. I hope it has been beneficial for you.
```

File: content/post/devsecops-azure-terraform-python.md (hunk 8 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:61]: Will fix

File: content/post/devsecops-azure-terraform-python.md (hunk 9 above)
Comment (RaphHaddad) [reply content/post/devsecops-azure-terraform-python.md:200]: Will fix. 
```suggestion
As well as a service connection from Azure DevOps into Azure Cloud that can be
```

**Comment by RaphHaddad** [comment]
Thanks @glav! I'll merge and I promise you another blog post to justify the technology choices of Python, Terraform, and the Azure CLI.
